### PR TITLE
Move To-Do Tool to Session Scope from Agent Scope

### PR DESCRIPTION
**Pull request by tlongwell-block**
## Overview
Refactors the TODO tool functionality from agent-scoped (shared across all sessions) to session-scoped (isolated per session). This change provides better data isolation, persistence, and user experience by ensuring each session maintains its own independent TODO list.

### Architecture
- **Removed** agent-level TODO storage (`Arc<Mutex<String>>` from Agent struct)
- **Added** session-level TODO storage (`todo_content: Option<String>` in SessionMetadata)
- **Implemented** session-aware TODO tool dispatch methods with proper error handling

### Session Storage
- Enhanced `SessionMetadata` with `todo_content` field
- Added backward-compatible deserializer for existing sessions
- TODO content now persists with session data in `.jsonl` files

### Tool Behavior
- TODO read/write operations now require active session context
- Character limits still enforced via `GOOSE_TODO_MAX_CHARS` environment variable
- Graceful error handling when no session context is available

### 🔒 **Data Isolation**
Each session maintains its own TODO list, preventing cross-contamination between different conversations or users.

### 💾 **Persistence** 
TODO lists survive agent restarts and are automatically saved/restored with session data.

### 🔄 **Backward Compatibility**
Existing sessions without TODO content continue to work seamlessly with empty TODO lists.

### 🧪 **Robust Testing**
Comprehensive integration test suite covering persistence, isolation, concurrency, and edge cases.

## Breaking Changes
- TODO tools now require session context and will error if called without it
- Agent-scoped TODO data is not migrated (sessions start with empty TODO lists)

## Files Changed
- `crates/goose/src/agents/agent.rs` - Core TODO dispatch logic
- `crates/goose/src/session/storage.rs` - Session metadata enhancement  
- `crates/goose/tests/todo_session_integration.rs` - New integration tests
- `crates/goose/tests/todo_tools_test.rs` - Removed (agent-scoped tests)
- Various test files updated for new metadata structure

## Testing
- ✅ Session isolation and persistence
- ✅ Character limit enforcement  
- ✅ Special character handling
- ✅ Concurrent access scenarios
- ✅ Backward compatibility
- ✅ Error handling edge cases

This change significantly improves the user experience by providing persistent, isolated TODO lists that follow session boundaries rather than being shared across the entire agent lifecycle.

### CODE REVIEW

File: crates/goose/src/agents/agent.rs [hunk 1]
```diff
@@ -285,20 +274,32 @@ impl Agent {
         }
     }
 
-    async fn handle_approved_and_denied_tools(
+    async fn handle_approved_and_denied_tools_with_session(
```
Comment (katzdave) [inline crates/goose/src/agents/agent.rs]: nit: Let's leave the name

File: crates/goose/src/agents/agent.rs [hunk 2]
```diff
@@ -285,20 +274,32 @@ impl Agent {
...
         for request in &permission_check_result.approved {
             if let Ok(tool_call) = request.tool_call.clone() {
-                let (req_id, tool_result) = self
-                    .dispatch_tool_call(tool_call, request.id.clone(), cancel_token.clone())
-                    .await;
+                let (req_id, tool_result) = if tool_call.name == TODO_READ_TOOL_NAME
+                    || tool_call.name == TODO_WRITE_TOOL_NAME
+                {
+                    // Handle TODO tools with session context
+                    self.dispatch_todo_tool_with_session_internal(
```
Comment (katzdave) [inline crates/goose/src/agents/agent.rs]: Any reason we can't just add the optional session to dispatch_tool_call, and only use it for TODO? Seems like a few extra functions got added here.

File: crates/goose/src/agents/agent.rs [hunk 3]
```diff
@@ -489,46 +645,21 @@ impl Agent {
                 None,
             )))
         } else if tool_call.name == TODO_READ_TOOL_NAME {
-            // Handle task planner read tool
-            let todo_content = self.todo_list.lock().await.clone();
-            ToolCallResult::from(Ok(vec![Content::text(todo_content)]))
+            // Handle task planner read tool - session-scoped
+            // This tool is now handled in dispatch_tool_call_with_session
+            ToolCallResult::from(Err(ErrorData::new(
```
Comment (katzdave) [inline crates/goose/src/agents/agent.rs]: Another weird consequence of the extra branching.

**Review by katzdave — COMMENTED** [review]
Looks good, did some testing! Love that now I can pick up my todo list when resuming a session, can we just reduce the branching.

One other thing I noticed is the TODO: READ tool doesn't return any data, which provides a poor experience when resuming a session. Is that an easy fix? If not can follow up with that later.

Also when you pull in main again, make sure to flip it back on, https://github.com/block/goose/pull/4158 disabled it.

File: crates/goose/src/agents/agent.rs (hunk 1 above)
Comment (tlongwell-block) [reply crates/goose/src/agents/agent.rs]: Done!

File: crates/goose/src/agents/agent.rs (hunk 2 above)
Comment (tlongwell-block) [reply crates/goose/src/agents/agent.rs]: Done

File: crates/goose/src/agents/agent.rs [hunk 4]
```diff
@@ -758,8 +798,7 @@ impl Agent {
             ]);
 
             // Add task planner tools
-            // TODO: Re-enable after next release
-            // prefixed_tools.extend([todo_read_tool(), todo_write_tool()]);
+            prefixed_tools.extend([todo_read_tool(), todo_write_tool()]);
```
Comment (katzdave) [inline crates/goose/src/agents/agent.rs:799]: 👍

File: crates/goose/src/session/storage.rs [hunk 5]
```diff
@@ -84,6 +86,7 @@ impl<'de> Deserialize<'de> for SessionMetadata {
             accumulated_input_tokens: Option<i32>,
             accumulated_output_tokens: Option<i32>,
             working_dir: Option<PathBuf>,
+            todo_content: Option<String>, // For backward compatibility
```
Comment (katzdave) [inline crates/goose/src/session/storage.rs:89]: Assuming this lets us load old sessions from before this feature?

**Review by katzdave — APPROVED** [review]
Nice job! Excited to have this back.

**Comment by tlongwell-block** [comment]
Oh boy. Goose borked this branch something fierce. I'll patch it up. Just a minute

**Comment by michaelneale** [comment]
lending my 👍  my support to this, as I really think we are missing out on this (I think that failure needs to run a `just` command to update the spec)

File: crates/goose/src/context_mgmt/auto_compact.rs [hunk 6]
```diff
@@ -441,23 +442,34 @@ mod tests {
...
+                 aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate \
+                 velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat \
+                 non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. \
+                 Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque \
+                 laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi \
+                 architecto beatae vitae dicta sunt explicabo. Nemo enim ipsam voluptatem quia voluptas \
+                 sit aspernatur aut odit aut fugit, sed quia consequuntur magni dolores eos qui ratione \
+                 voluptatem sequi nesciunt. Neque porro quisquam est, qui dolorem ipsum quia dolor sit \
+                 amet, consectetur, adipisci velit, sed quia non numquam eius modi tempora incidunt ut \
+                 labore et dolore magnam aliquam quaerat voluptatem.",
```
Comment (DOsinga) [inline crates/goose/src/context_mgmt/auto_compact.rs]: why these changes?

Comment (tlongwell-block) [reply crates/goose/src/context_mgmt/auto_compact.rs]: The test was flakey for me and this fixed that

Comment (tlongwell-block) [reply crates/goose/src/context_mgmt/auto_compact.rs]: cc @katzdave

Comment (DOsinga) [reply crates/goose/src/context_mgmt/auto_compact.rs]: ok, but in what way was it flaky? I just don't see how this changes the behavior, other than reducing the number of tokens we need and increasing the number of tokens we send. if that is flakey, it really means that something underlying must be broken

Comment (tlongwell-block) [reply crates/goose/src/context_mgmt/auto_compact.rs]: reverted

File: crates/goose/src/session/storage.rs (hunk 5 above)
Comment (tlongwell-block) [reply crates/goose/src/session/storage.rs:89]: Yes!
